docs/tutorial_history_backend.rst

```diff
@@ -0,0 +1,319 @@
+            msg = 'error when query db: {}: {}'.format(e.__class__.__name__, e)
+            print(msg, file=sys.stderr)
+            return
+        data = json.loads(r.text)
+        for item in data['rows']:
+            cmd = item['doc'].copy()
+            cmd['ts'] = cmd['ts'][0]
+            yield cmd
+
+The `try-except` is here so that we're safe when something bad happened, like
+couchdb is not get started, etc.
+
+
+Try Out Our New History Backend
+===============================
+
+That's it. Your can find full code here:
+`<https://gist.github.com/mitnk/2d08dc60aab33d8b8b758c544b37d570>`_
+
+Let's start a new xonsh session:
+
+.. code-block:: none
+
+    $ history info
+    backend: couchdb
+    sessionid: 1486035364166-3bb78606-dd59-4679
+
+    $ ls
+    Applications   Desktop    Documents    Downloads
+
+    $ echo hi
+    hi
+
+Start a second xonsh session:
+
+.. code-block:: none
+
+    $ history info
+    backend: couchdb
+    sessionid: 1486035430658-6f81cd5d-b6d4-4f6a
+
+    $ echo new
+    new
+
+    $ history show all -nt
+    0:(2017-02-02 19:36) history info
+    1:(2017-02-02 19:36) ls
+    2:(2017-02-02 19:37) echo hi
+    3:(2017-02-02 19:37) history info
+    4:(2017-02-02 19:37) echo new
+
+    $ history -nt
+    0:(2017-02-02 19:37) history info
+    1:(2017-02-02 19:37) echo new
+    2:(2017-02-02 19:37) history show all -nt
+
+We don't miss any histories, so we're good I think.
+
+
+History Garbage Collection
+==========================
+
+In built-in history backends ``json``, ``sqlite``, GC will happen when
+xonsh get started or when run command ``history gc``. History items that
+range out of what `$XONSH_HISTORY_SIZE <envvars.html#xonsh-history-size>`_
+defines will be deleted.
+
+.. code-block:: python
+
+    class History:
+        def run_gc(self, size=None, blocking=True):
+            """Run the garbage collector.
+
+            Parameters
+            ----------
+            size: None or tuple of a int and a string
+                Detemines the size and units of what would be allowed to remain.
+            blocking: bool
+                If set blocking, then wait until gc action finished.
+            """
+            pass
+
+The History public method ``run_gc()`` is for this purpose. Our
+``CouchDBHistory`` define this method, thus it inherits from its parent
+`History`, which does nothing. We will leave the GC implementing as an
+exercise.
+
+
+Other History Options
+=====================
+
+There are some environment variables that could change the behaviors of
+history backend. Such as `$HISTCONTROL <envvars.html#histcontrol>`_,
+`$XONSH_HISTORY_SIZE <envvars.html#xonsh-history-size>`_,
+`$XONSH_STORE_STDOUT <envvars.html#xonsh-store-stdout>`_, etc.
+
+We should implement these ENVs in our CouchDB backend. Luckily, it's not a
+hard thing. We will leave these features implementing for yourself.
+
+
+Wrap Up
+=======
+
+Though the code are written as a just-work-level. But it does show us
```

Comment:
This is a barebones implementation but hopefully it will give you a sense of how you can customize xonsh's history backend for your own needs!